### DISCUSSION TITLE
God-oriented programming meaning and is it an anti-pattern

### DISCUSSION THREAD

**Discussion by Ivansstyle** [idealvin/coost#343] [General]
Hi coost developers, 

Have never came across this idea of god-oriented programming, and decided to research it a bit. 
Per [Wikipedia](https://en.wikipedia.org/wiki/God_object) God object is an anti-pattern from a software engineering perspective, however there is question if God-oriented programming and God-object are related. 

Could you please comment on what this should mean and how to understand this?

**Comment by 909254** [comment]
God said, "Let there be light." And there was light.
God said, "This project has no bugs." And the project had no bugs.

**Comment by idealvin** [comment]
hi, @Ivansstyle 

`god-oriented` is not `object-oriented`,  it has no relationship with the `god-object`. Instead, it is related to templates.

In C++, abuse of templates will reduce the readability of the code. Some code written with templates were thought to be “god-only-knowns”. We jokingly call it god-oriented programming.

In Chinese, we call god-oriented `面向玄学`. In the Chinese concept, god refers to `老天`, but not an almighty god.

Coost provides some template-related features in `co/god.h`, perhaps the most interesting one is `god::bless_no_bugs()`.
